Convert constraints from Prolog to JavaScript

mcmire opened the pull request:
## Explanation

<!--
Thanks for your contribution! Take a moment to answer these questions so that reviewers have the information they need to properly understand your changes:

* What is the current state of things and why does it need to change?
* What is the solution your changes offer and how does it work?
* Are there any changes whose purpose might not obvious to those unfamiliar with the domain?
* If your primary goal was to update one package but you found you had to update another one along the way, why did you do so?
* If you had to upgrade a dependency, why did you do so?
-->

The existing Yarn constraints have been difficult to maintain because they are written in Prolog. More recent versions of Yarn support JavaScript-based constraints, so this commit rewrites the existing constraints in JavaScript.

Instead of `constraints.pro`, all constraints are now kept in `yarn.config.cjs`. This file can be used for other things in the future besides constraints. (And constraints can be used in the future to check more than just dependencies!)

Note that we have had discussions in the past around the peer dependency constraints, and we know that they are not quite correct. This will be fixed in another PR.

## References

<!--
Are there any issues that this pull request is tied to? Are there other links that reviewers should consult to understand these changes better?

For example:

* Fixes #12345
* Related to #67890
-->

Fixes #4384.

## Changelog

<!--
If you're making any consumer-facing changes, list those changes here as if you were updating a changelog, using the template below as a guide.

(CATEGORY is one of BREAKING, ADDED, CHANGED, DEPRECATED, REMOVED, or FIXED. For security-related issues, follow the Security Advisory process.)

Please take care to name the exact pieces of the API you've added or changed (e.g. types, interfaces, functions, or methods).

If there are any breaking changes, make sure to offer a solution for consumers to follow once they upgrade to the changes.

Finally, if you're only making changes to development scripts or tests, you may replace the template below with "None".
-->

(No consumer-facing changes.)

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

mcmire commented on yarn.config.cjs:
Make sure to review `yarn.config.cjs` :) It is being hidden in the diff.

cryptodev-2s commented on yarn.config.cjs:
```diff
@@ -0,0 +1,654 @@
+
+        // All non-root package must have a valid "changelog:validate" script.
+        expectCorrectWorkspaceChangelogValidationScript(workspace);
+
```
A rule to check `changelog:update` script exist is missing

mcmire replied:
Hmm, you're right. This is something we didn't have before, but I can add it. I've added that here: [d408971](https://github.com/MetaMask/core/pull/4546/commits/d408971400d72377c2c451bfa0d1e6fa85054d5d)

cryptodev-2s commented on yarn.config.cjs:
```diff
@@ -0,0 +1,654 @@
+          workspace.manifest.license === undefined
+        ) {
+          expectWorkspaceField(workspace, 'license', 'MIT');
+        }
```
the following rule is missing for the packages that are under ISC license
```
% The license for all published packages must be MIT unless otherwise specified.
gen_enforced_field(WorkspaceCwd, 'license', 'MIT') :-
  \+ workspace_field(WorkspaceCwd, 'private', true),
  WorkspaceCwd \= 'packages/json-rpc-engine',
  WorkspaceCwd \= 'packages/json-rpc-middleware-stream',
  WorkspaceCwd \= 'packages/permission-log-controller',
  WorkspaceCwd \= 'packages/eth-json-rpc-provider'.
% The following published packages use an ISC license instead of MIT.
gen_enforced_field(WorkspaceCwd, 'license', 'ISC') :-
  \+ workspace_field(WorkspaceCwd, 'private', true),
  (
    WorkspaceCwd == 'packages/json-rpc-engine' ;
    WorkspaceCwd == 'packages/json-rpc-middleware-stream' ;
    WorkspaceCwd == 'packages/eth-json-rpc-provider'
  ).
```

mcmire replied:
I decided to update these checks slightly so that we don't have to keep track of which packages _aren't_ using MIT. My thought here was that if someone is adding a new package and forgets to set the license, the constraints will ask them to use MIT. However, if a package is already using a license, then that should be okay, and it can use whatever license it wants to. Does that make sense or do you think it is still worth it to track those exceptions here?

cryptodev-2s replied:
I see your point. My only concern is that this could allow someone to unintentionally commit a package under a license we do not want to permit.

mcmire replied:
I've updated this check to enforce MIT if `license` is missing, but to also make an exception for `json-rpc-engine`, `json-rpc-middleware-stream`, `permission-log-controller`, and `eth-json-rpc-provider`. I also realized that I could make sure that the LICENSE file was present, so I added a check for that and extracted the whole check to a function to clean up the code a bit more: [c51cc83](https://github.com/MetaMask/core/pull/4546/commits/c51cc83774c3dc510c022f5db7d3373d5c33bf6a)

cryptodev-2s commented:
Looks good to me. However, there are two rules that are missing which it will be nice if we can add them. Additionally, it would be helpful to create a ticket to address the inconsistency with some rules. For instance, we check that `changelog:validate` has the correct value, but if it doesn't exist, no error is thrown. Same for `changelog:update` and others...

cryptodev-2s commented:
We should add a summary of the rules we are checking at the top of the file, like this:
```
/**
 * Yarn Workspace Constraints Configuration
 *
 * This configuration ensures that all workspaces in the monorepo adhere to a consistent set of rules and conventions.
 *
 * Rules:
 * 
 * 1. **Workspace Name**:
 *    - All packages must have a name.
 *    - Non-root packages must have a name that matches their directory (`@metamask/{dirname}`).
 *
 * 2. **Package Version**:
 *    - All non-root packages must have a version.
 *
 * 3. **Package Description**:
 *    - All packages must have a description that does not end with a period.
 *
 * 4. **NPM Keywords**:
 *    - Non-root packages must have the keywords `MetaMask` and `Ethereum`.
 *
 * 5. **Homepage URL**:
 *    - Non-root packages must have a homepage URL that includes their name and points to the correct repository.
 *
 * 6. **Bugs URL**:
 *    - Non-root packages must have a bugs URL pointing to the Issues page of the repository.
 *
 * 7. **Repository**:
 *    - All packages must specify Git as the repository type.
 *    - The repository URL must point to a MetaMask GitHub repository.
 *
 * 8. **License**:
 *    - Default license for non-root packages is MIT, except for specific packages (`ISC` or custom).
 *
 * 9. **Side Effects**:
 *    - Non-root packages must not have side effects, except `@metamask/base-controller`.
 *
 * 10. **Exports**:
 *     - Non-root packages must set up ESM- and CommonJS-compatible exports correctly.
 *
 * 11. **Build and Test Scripts**:
 *     - Non-root packages must have the same `build`, `build:docs`, `test`, `test:clean`, `test:verbose`, and `test:watch` scripts.
 *     - Non-root packages must not have a `prepack` script.
 *     - Non-root packages must have a valid `changelog:validate` script.
 *     - Non-root packages must have a valid `changelog:update` script.
 *
 * 12. **Included Files**:
 *     - Non-root packages must only include files generated during the build process.
 *     - Root package must specify an empty set of published files.
 *
 * 13. **Yarn Version**:
 *     - The root workspace must specify the Yarn version required for development.
 *
 * 14. **Node.js Version**:
 *     - All packages must specify a minimum Node.js version of 18.18.
 *
 * 15. **Publishing**:
 *     - Non-root public packages should be published to the NPM registry.
 *     - Non-root private packages should not be published.
 *
 * 16. **README.md**:
 *     - Non-root packages must have a valid README.md file.
 *
 * 17. **Dependency Constraints**:
 *     - Dependencies and devDependencies must match the current version of the dependency.
 *     - No dependency should be listed under both dependencies and devDependencies.
 *     - Controller packages must also be listed in peerDependencies.
 *     - Version ranges for the same dependency must be consistent across the monorepo.
 *
 * This configuration is enforced by defining rules and validations within the Yarn constraints.
 */
```

mcmire commented:
@cryptodev-2s We could add a summary at the top, but I'm already summarizing each step, and then on top of that I have documentation for each function that explains how certain constraints work. So if we added a summary, that would be a third form of documentation that people would have to remember to update if they updated the constraints. I guess we can regenerate it easily using ChatGPT, but maybe there's another way we can do this? If I extracted every constraint to its own function, would that help?

cryptodev-2s reviewed (approved):
LGTM!
